### PR TITLE
Feat/verificator2

### PR DESCRIPTION
**Pull request by matiu**
Adds verification to server given address.

### CODE REVIEW

File: bit-wallet/pepe [hunk 1]
```diff
@@ -0,0 +1 @@
+{"m":1,"n":1,"walletPrivKey":"{\"bn\":\"54ca448f1c56ffe7f1681f1cdcd398327516de2e0dcde42875dcd2324fb4cf5\",\"compressed\":true,\"network\":\"livenet\"}","network":"testnet","xPrivKey":"tprv8ZgxMBicQKsPczBpizHdcEX8dMNu2ohYHS3mFoNetGzFWARJVcZt7UtsEcar64qBkz2Jw7uBobU7XNh8GNssVfiepnCNyMa6ymNmGMXZN76","copayerId":"cefc16bb-a4c9-4c82-b055-e21afc5c6f35","signingPrivKey":"ef21d7c9d7e96c91d082d085bc4c1f77b62d8dd6b6c602473155ba56276ab92a","publicKeyRing":["tpubD6NzVbkrYhZ4WTDccdxE1eBFCNtqC8tSrjeYYKQxJYneLeg581PUHyWjQkxfnsvuNMESuPDSWBpHhzZCbJFrjto51ttsWE4S7EvatTc7W3C"]}
```
Comment (isocolsky) [inline bit-wallet/pepe]: this was eventually going to end up in here lol

**Comment by matiu** [comment]
updated.

File: lib/client/Verifier.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+var Bitcore = require('bitcore');
+var BitcoinUtils = require('../bitcoinutils')
+
+function Verifier(opts) {};
+
+Verifier.checkAddress = function(data, address) {
+    var local =  BitcoinUtils.deriveAddress(data.publicKeyRing, address.path, data.m, data.network);
+    if (local.address != address.address || JSON.stringify(local.publicKeys)!= JSON.stringify(address.publicKeys))
+      return cb('Server sent a fake address.');
```
Comment (isocolsky) [inline lib/client/Verifier.js]: there is no cb()

File: lib/client/api.js [hunk 3]
```diff
@@ -281,7 +283,16 @@ API.prototype.createAddress = function(cb) {
...
 
   var url = '/v1/addresses/';
-  this._doPostRequest(url, {}, data, cb);
+  this._doPostRequest(url, {}, data, function(err, address) {
+    if (err) return cb(err);
+
+    if (data.publicKeyRing.length != data.n)
+      return cb('Wallet Incomplete, cannot derive address.')
+
+    Verifier.checkAddress(data, address);
```
Comment (isocolsky) [inline lib/client/api.js]: should be `if (Verifier.checkAddress(data, address)) return cb('Server sent fake address');`

**Comment by matiu** [comment]
updated

File: lib/client/api.js [hunk 4]
```diff
@@ -9,6 +9,8 @@ log.debug = log.verbose;
 
 var Bitcore = require('bitcore')
 var SignUtils = require('../signutils');
+var Verifier = require('./verifier');
+var ServerCompromissedError = require('./servercompromissederror')
```
Comment (isocolsky) [inline lib/client/api.js]: I think the correct spelling is compromised

File: lib/client/api.js [hunk 5]
```diff
@@ -281,7 +286,18 @@ API.prototype.createAddress = function(cb) {
   var data = this._loadAndCheck();
 
   var url = '/v1/addresses/';
-  this._doPostRequest(url, {}, data, cb);
+  this._doPostRequest(url, {}, data, function(err, address) {
+    if (err) return cb(err);
+
+    if (data.publicKeyRing.length != data.n)
```
Comment (isocolsky) [inline lib/client/api.js]: I'm not sure this is correct. Shouldn't this check be performed server-side only?

**Comment by matiu** [comment]
updated.

**Comment by matiu** [comment]
updated.

**Pull request merged**